staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_discovery.go

```diff
@@ -117,6 +125,72 @@ func (r *groupDiscoveryHandler) unsetDiscovery(group string) {
 	delete(r.discovery, group)
 }
 
+// Groups returns the list of API groups that are served by the API extensions server.
+// It uses the CustomResourceDefinitionLister to determine which groups are served.
+func (r *groupDiscoveryHandler) Groups(ctx context.Context, _ *http.Request) ([]metav1.APIGroup, error) {
```

Comment:
I'm a bit confused why we need to implement this from scratch. apiextensions should have discovery information served, is that not accessible here?

Reply:
Its implemented but not in a convenient form - served from cache which is populated from controller spinning and monitoring crds. All this is based on handlers, so not quite easy to consume programmatically. 

This change kinda unblocks controlplanes work, as it makes to consume extension server content easier via easier route. But totally agree, confusing :D and sorry for this.

As discussed this needs a wider view and maybe some architecture documents. Will try to find some time to kick something off.

Reply:
I re-did the implementation using existing machinery.

Reply:
Question (also maybe @sttts might know) based on the interface defined in https://github.com/kubernetes/kubernetes/pull/127524, it looks like we need to pass in a context and empty request to satisfy the interface and fill in ServerAddressByClientCIDRs. https://github.com/mjudeikis/kubernetes/blob/ee55200440c8236248f47cbe2dd783ba1a717614/staging/src/k8s.io/apiserver/pkg/endpoints/discovery/root.go#L128. We don't exactly need this right? It seems a bit weird that we need the two extra parameters in the interface that are not being used.

Reply:
In general it was due to fact I added additional method to webserver :) kinda looked natural to add ctx and request into webserver methods :)